logstash_async/memory_cache.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+#
+# This software may be modified and distributed under the terms
+# of the MIT license.  See the LICENSE file for details.
+
+import uuid
+from datetime import datetime, timedelta
+from logstash_async.cache import Cache
+
+
+class MemoryCache(Cache):
+    """Backend implementation for python-logstash-async. Keeps messages in a local, in-memory cache
+    while attempting to publish them to logstash. Does not persist through process restarts. Also,
+    does not write to disk.
+
+    :param cache: Usually just an empty dictionary
+    :param event_ttl: Optional parameter used to expire events in the cache after a time
+    """
+
+    def __init__(self, cache, event_ttl=None):
+        self._cache = cache
+        self._event_ttl = event_ttl
+
+    def add_event(self, event):
+        event_id = uuid.uuid4()
+        self._cache[event_id] = {
+            "event_text": event,
+            "pending_delete": False,
+            "entry_date": datetime.now(),
+            "id": event_id
+        }
+
+    def get_queued_events(self):
+        events = []
+        for event in self._cache.values():
+            if not event['pending_delete']:
+                events.append(event)
+        return events
+
+    def requeue_queued_events(self, events):
+        for event in events:
+            self._cache[event['id']]['pending_delete'] = False
```

Comment:
While not very likely, maybe a try-except would be useful here if `self._cache` does not contain `event['id']`. Just to prevent breaking the whole process and instead log a warning or so.

Reply:
Hello. I'm happy to implement these changes, but I have a question about "logging the warning". I'm not exactly sure how we should be logging a warning. We could print to STDOUT if you'd like, but using a logger would probably just end up in another enqueued message, which could cause this problem again.

What are your thoughts?

Reply:
For other errors, the logging framework is used as well even if the logged message could not be sent directly (e.g. on network errors). But using the logging framework enables other handlers as well to be fired (e.g. a mail handler, a stderr handler or anything else). So even if we are unable to send our events, other handlers might inform the user about the problem.

We have `LogProcessingWorker._safe_log` to more or less safely log a message: use the logging framework except we are in the process of a shutdown, then just print to `stderr`.
We could move this method into `utils.py` to make it a more easy accesible function.

In general, I don't like basic `stderr` (or `stdout`) logging very much. I often work on projects where there is no `stderr`/`stdout` because the application is a daemon. IMO it should be just a last resort if everything else failed.